storage/innobase/include/trx0purge.h

```diff
@@ -58,76 +58,64 @@ ulint trx_purge(ulint n_tasks, ulint history_size);
-    struct trx_rseg_cmp
-    {
-      /** Compare two trx_rseg_t* based on trx_no.
-      @param lhs first index in trx_sys.rseg_array to compare
-      @param rhs second index in trx_sys.rseg_array to compare
-      @return whether lhs>rhs */
-      bool operator()(const byte lhs, const byte rhs)
-      {
-        ut_ad(lhs < TRX_SYS_N_RSEGS);
-        ut_ad(rhs < TRX_SYS_N_RSEGS);
-        /* We can compare without trx_rseg_t::latch, because rseg last
-        commit is always set before pushing rseg to purge queue. */
-        return trx_sys.rseg_array[lhs].last_commit_and_offset >
-               trx_sys.rseg_array[rhs].last_commit_and_offset;
-      }
-    };
-    byte *begin() { return m_array; }
-    byte *end() { return m_end; }
-    const byte *c_begin() const { return m_array; }
-    const byte *c_end() const { return m_end; }
-    size_t size() const
-    {
-      size_t s= c_end() - c_begin();
-      ut_ad(s <= TRX_SYS_N_RSEGS);
-      return s;
-    }
-    bool empty() const { return !size(); }
-    void clear() { m_end= m_array; }
+    typedef std::vector<uint64_t, ut_allocator<uint64_t>> container_type;
+
+    bool empty() const { return m_array.empty(); }
+    void clear() { m_array.clear(); }
+    container_type::const_iterator cbegin() const { return m_array.cbegin(); }
+    container_type::const_iterator cend() const { return m_array.end(); }
 
-    /** Push index of trx_sys.rseg_array into min-heap.
-    @param i index to push */
-    void push_rseg_index(byte i)
+    /** Push (trx_no, trx_sys.rseg_array index) into min-heap.
+    @param trx_no_rseg (trx_no << 8 | (trx_sys.rseg_array index)) */
+    void push_trx_no_rseg(container_type::value_type trx_no_rseg)
     {
-      ut_ad(i < TRX_SYS_N_RSEGS);
-      ut_ad(size() + 1 <= TRX_SYS_N_RSEGS);
-      *m_end++= i;
-      std::push_heap(begin(), end(), trx_rseg_cmp());
+      m_array.push_back(trx_no_rseg);
+      std::push_heap(m_array.begin(), m_array.end(),
+                     std::greater<container_type::value_type>());
     }
 
     /** Push rseg to priority queue.
-    @param rseg trx_rseg_t pointer to push */
-    void push(const trx_rseg_t *rseg)
+    @param trx_no trx_no of committed transaction
+    @param rseg rseg of committed transaction*/
+    void push(trx_id_t trx_no, const trx_rseg_t &rseg)
     {
-      ut_ad(rseg >= trx_sys.rseg_array);
-      ut_ad(rseg < trx_sys.rseg_array + TRX_SYS_N_RSEGS);
-      byte i= byte(rseg - trx_sys.rseg_array);
-      push_rseg_index(i);
+      ut_ad(&rseg >= trx_sys.rseg_array);
+      ut_ad(&rseg < trx_sys.rseg_array + TRX_SYS_N_RSEGS);
+      push_trx_no_rseg(trx_no << 8 | byte(&rseg - trx_sys.rseg_array));
+    }
+
+    /** Extracts rseg from (trx_no, trx_sys.rseg_array index) pair.
+    @param trx_no_rseg (trx_no << 8 | (trx_sys.rseg_array index)
+    @return pointer to rseg in trx_sys.rseg_array */
+    static trx_rseg_t *rseg(container_type::value_type trx_no_rseg) {
+      byte i= trx_no_rseg & ((1ULL << 8) - 1);
```

Comment:
This would be written simpler just `byte i= byte(trx_no_rseg);`

I would not repeat the `trx_no << 8 | …` in multiple places. Maybe, introduce `static constexpr unsigned TRX_NO_SHIFT= 8` and use it where applicable. It might be more efficient to reserve 16 bits for the 7-bit index: I’d compare the generated code on IA-32.

Reply:
Both 8 and 16 bit shifts are compiled to  `shld` and `shl` instructions for IA-32:
```
0x80ae1259 <trx_t::commit_low(mtr_t*)+1865>     shld   $0x8,%ecx,%ebx           
0x80ae125d <trx_t::commit_low(mtr_t*)+1869>     shl    $0x8,%ecx
```